projects/baby_driver/test/test_dispatcher.c

```diff
@@ -0,0 +1,126 @@
+#include "dispatcher.h"
+
+#include <string.h>
+
+#include "babydriver_msg_defs.h"
+#include "can.h"
+#include "can_msg_defs.h"
+#include "can_transmit.h"
+#include "log.h"
+#include "ms_test_helper_can.h"
+#include "ms_test_helpers.h"
+#include "test_helpers.h"
+#include "unity.h"
+
+#define TEST_BABYDRIVER_MESSAGE_ID (NUM_BABYDRIVER_MESSAGES - 1)
+
+typedef enum {
+  TEST_CAN_EVENT_TX = 0,
+  TEST_CAN_EVENT_RX,
+  TEST_CAN_EVENT_FAULT,
+  NUM_TEST_CAN_EVENTS,
+} TestCanEvent;
+
+static CanStorage s_can_storage;
+
+static uint8_t s_times_callback_called;
+static uint8_t s_received_data[8];
+static void *s_received_context;
+
+static bool s_should_tx_result;
+static StatusCode s_status_return;
+
+static StatusCode prv_dispatcher_callback(uint8_t data[8], void *context, bool *tx_result) {
+  s_times_callback_called++;
+  memcpy(s_received_data, data, 8);
+  s_received_context = context;
+  *tx_result = s_should_tx_result;
+  return s_status_return;
+}
+
+void setup_test(void) {
+  initialize_can_and_dependencies(&s_can_storage, SYSTEM_CAN_DEVICE_BABYDRIVER, TEST_CAN_EVENT_TX,
+                                  TEST_CAN_EVENT_RX, TEST_CAN_EVENT_FAULT);
+  TEST_ASSERT_OK(dispatcher_init());
+
+  s_times_callback_called = 0;
+  memset(s_received_data, 0, sizeof(s_received_data));
+  s_received_context = NULL;
+  s_should_tx_result = false;
+  s_status_return = STATUS_CODE_OK;
+}
+void teardown_test(void) {}
+
+// Test that we can successfully dispatch.
+void test_dispatching_no_tx_result(void) {
+  uint8_t arbitrary_context;  // value unused, but serves as test context for the callback
+  TEST_ASSERT_OK(dispatcher_register_callback(TEST_BABYDRIVER_MESSAGE_ID, prv_dispatcher_callback,
+                                              &arbitrary_context));
+  TEST_ASSERT_EQUAL(0, s_times_callback_called);
+
+  // send CAN message with our message ID - should receive same data
+  uint8_t data[7] = { 1, 2, 3, 4, 5, 6, 7 };
+  CAN_TRANSMIT_BABYDRIVER(TEST_BABYDRIVER_MESSAGE_ID, data[0], data[1], data[2], data[3], data[4],
+                          data[5], data[6]);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(1, s_times_callback_called);
+  TEST_ASSERT_EQUAL(TEST_BABYDRIVER_MESSAGE_ID, s_received_data[0]);
+  TEST_ASSERT_EQUAL(&arbitrary_context, s_received_context);
+
+  // use a pointer to the second element to compare |data| with the non-ID part of |s_received_data|
+  uint8_t *received_data = &s_received_data[1];
+  TEST_ASSERT_EQUAL_UINT8_ARRAY(data, received_data, 7);
+
+  // should not tx result, so no CAN events
```

Comment:
nit: should explicitly set s_should_tx_result for more clarity (I had to scroll back up to understand why it shouldn’t to result) but also I’m just lazy